File: doughmination.win/backend/app/utils/cache.py
"""
Simple in-memory caching utility
"""

import time
from typing import Any, Optional

_cache = {}

def get_from_cache(key: str) -> Optional[Any]:
    """
    Get a value from cache if it exists and hasn't expired
    
    Args:
        key: Cache key
    
    Returns:
        Cached value or None if expired/not found
    """
    if key in _cache:
        value, expire_time = _cache[key]
        if time.time() < expire_time:
            return value
        else:
            del _cache[key]  # Clean up expired entry
    return None

def set_in_cache(key: str, value: Any, ttl: int = 30):
    """
    Set a value in cache with TTL
    
    Args:
        key: Cache key
        value: Value to cache
        ttl: Time to live in seconds (default: 30)
    """
    expire_time = time.time() + ttl
    _cache[key] = (value, expire_time)

def clear_cache(key: Optional[str] = None):
    """
    Clear cache entry or entire cache
    
    Args:
        key: Specific key to clear, or None to clear all
    """
    if key:
        _cache.pop(key, None)
    else:
        _cache.clear()

File: co/system/backend/src/utils/cache.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Simple in-memory caching utility.
 */

interface CacheEntry {
  value: unknown;
  expireAt: number; // epoch ms
}

const cache = new Map<string, CacheEntry>();

/**
 * Get a value from cache if it exists and hasn't expired.
 *
 * @param key - Cache key
 * @returns Cached value, or undefined if expired/not found
 */
export function getFromCache<T = unknown>(key: string): T | undefined {
  const entry = cache.get(key);
  if (entry) {
    if (Date.now() < entry.expireAt) {
      return entry.value as T;
    }
    cache.delete(key); // Clean up expired entry
  }
  return undefined;
}

/**
 * Set a value in cache with a TTL.
 *
 * @param key - Cache key
 * @param value - Value to cache
 * @param ttlSeconds - Time to live in seconds (default: 30)
 */
export function setInCache(key: string, value: unknown, ttlSeconds = 30): void {
  cache.set(key, { value, expireAt: Date.now() + ttlSeconds * 1000 });
}

/**
 * Clear a cache entry, or the entire cache.
 *
 * @param key - Specific key to clear, or omit to clear all
 */
export function clearCache(key?: string): void {
  if (key) {
    cache.delete(key);
  } else {
    cache.clear();
  }
}